cafebabel/static/js/articles-tags.js

```diff
@@ -0,0 +1,180 @@
+      )
+    )
+
+    return container
+  }
+  _createTag(tagValue, index) {
+    return `
+      <li>
+        ${tagValue}
+        <input name="tag-${++index}" list="tags" value=${tagValue} type="hidden">
+        <button></button>
+      </li>
+    `
+  }
+  _inactiveSuggestionsList() {
+    this.suggestions.classList.add('inactive')
+  }
+  _activeSuggestionsList() {
+    this.suggestions.classList.remove('inactive')
+  }
+  _emptySuggestions() {
+    this.suggestions.innerHTML = ''
+  }
+  _emptyAddField() {
+    this.fieldAdd.value = ''
+  }
+  _renderSuggestion(container) {
+    this.suggestions.replaceWith(container)
+    TagEventListener.addSuggestion()
+  }
+  _render() {
+    const container = this.list.cloneNode(false)
+    this.list.replaceWith(this._createTagsList(container, this.values))
+    TagEventListener.addRemoveListener()
+    TagEffect.fadeIn(container.querySelector('li:last-child'))
+  }
+}
+
+class TagEffect {
+  static fadeIn(element) {
+    function fade() {
+      element.style.opacity = +element.style.opacity + 0.03
+      if (element.style.opacity <= 1) {
+        requestAnimationFrame(fade)
+      }
+    }
+    element.style.opacity = 0
+    fade()
+  }
+  static fadeOut(element) {
+    function fade() {
+      element.style.opacity = +element.style.opacity - 0.03
+      if (element.style.opacity < 0) {
+        element.style.display = 'none'
+      } else {
+        requestAnimationFrame(fade)
+      }
+    }
+    element.style.opacity = 1
+    fade()
+  }
+}
+
+class TagEventListener {
+  static clickAdd() {
+    const tags = new Tags()
+    tags.buttonAdd.addEventListener('click', event => {
+      event.preventDefault()
+      const submission = event.target.previousSibling.value
+      tags.addNewTag(submission)
+    })
+  }
+  static addSuggestion() {
+    const tags = new Tags()
+    tags.suggestions.querySelectorAll('li').forEach(li =>
+      li.addEventListener('click', event => {
+        const submission = event.target.innerText
+        tags.addNewTag(submission)
+      })
+    )
+  }
+  static addRemoveListener() {
+    const tags = new Tags()
+    tags.removeButtons.forEach(button =>
+      button.addEventListener('click', event => {
+        event.preventDefault()
+        const tagElement = event.target.parentNode
+        tags.removeTag(tagElement)
+      })
+    )
+  }
+  static keyup() {
+    const inputNewTag = document.querySelector('.tags input[name=tag-new]')
+    inputNewTag.addEventListener('keypress', event => {
+      /* Return if arrow up, arrow down or enter are pressed */
+      if (event.keyCode == 13) {
+        return
+      }
+      if (event.keyCode == 40 || event.keyCode == 38) return
+      const submission = event.target.value
+      if (submission.length < 3) return
+      const tags = new Tags()
+      request(
+        `/article/tag/suggest/?language=${tags.language}&terms=${submission}`
+      )
+        .then(tagsApi => tags.handleSuggestion(tagsApi))
```

Comment:
Can be simplified as `.then(tags.handleSuggestion)`

Reply:
@davidbgk - I feel like I'm losing `this` and the argument?

Reply:
Ah! Maybe not that simple 👌